### PR TITLE
Sql Cache record update fix + test (#2231)

### PR DESCRIPTION
**Pull request by PierLucaLW**
#2231

Not sure if you wanna still keep the _id field, in that case we should retreive it from the old record and pass it when updating it

### CODE REVIEW

**Comment by tasomaniac** [comment]
This approach will need migration. You've mentioned another potential solution in the code. Should we go for that? Looks much safer.

**Comment by PierLucaLW** [comment]
At first I thought the problem was there but that part was correct since oldRecord var gets its fields updated after `.mergeWith()`. The `UPDATE` instruction doesn't seem to work on it and `INSERT OR REPLACE` looks like it needs all the field, or at least the id since is a primary key. I can try getting the id from the old record and pass it when to update, would that be good?

File: apollo-normalized-cache-sqlite/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/sql/SqlNormalizedCache.kt [hunk 1]
```diff
@@ -9,89 +9,90 @@ import com.apollographql.apollo.cache.normalized.RecordFieldJsonAdapter
...
-      null
+    fun selectRecordForKey(key: String): Record? {
+        return try {
+            cacheQueries.recordForKey(key)
+                    .executeAsList()
+                    .firstOrNull()
+                    ?.let {
+                        Record.builder(it.key)
+                                .addFields(recordFieldAdapter.from(it.record)!!)
+                                .addField("_id", it._id)
```
Comment (PierLucaLW) [inline apollo-normalized-cache-sqlite/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/sql/SqlNormalizedCache.kt:70]: you might want to pass this in a nicer way, just showing it works here

File: apollo-normalized-cache-sqlite/src/commonTest/kotlin/com/apollographql/apollo/cache/normalized/sql/SqlNormalizedCacheTest.kt [hunk 2]
```diff
@@ -54,7 +54,19 @@ class SqlNormalizedCacheTest {
   }
 
   @Test
-  fun testRecordMerge() {
+  fun testRecordMerge_noOldRecord() {
```
Comment (sav007) [inline apollo-normalized-cache-sqlite/src/commonTest/kotlin/com/apollographql/apollo/cache/normalized/sql/SqlNormalizedCacheTest.kt:57]: I think you missing `createRecord(STANDARD_KEY)` before `cache.merge(`

**Comment by sav007** [comment]
Also could you please fix formatting

File: apollo-normalized-cache-sqlite/src/commonTest/kotlin/com/apollographql/apollo/cache/normalized/sql/SqlNormalizedCacheTest.kt (hunk 2 above)
Comment (PierLucaLW) [reply apollo-normalized-cache-sqlite/src/commonTest/kotlin/com/apollographql/apollo/cache/normalized/sql/SqlNormalizedCacheTest.kt:57]: I've split the test in 2 for when a record is already present or not to test both branches

**Comment by tasomaniac** [comment]
I'm still not sure if this is the right way. 1.4.x versions or any older do not replace into existing. So, I'm not sure why we would need a behavior change here.

The solution to the bug must be in the code. There must be something wrong with Kotlin migration.

**Comment by sav007** [comment]
The changes in this fix look legit, but I agree with @tasomaniac before merging we have to find why it was working up to 1.4.5 version. As we might be missing bigger issue with migrating SQL module to Kotlin.

Going to try 1.4.4 version to run the test.

**Comment by sav007** [comment]
🤦 🤦 🤦 🤦 🤦 🤦 
The issue is 🥁  🥁  🥁  🥁  🥁  🥁 
https://github.com/apollographql/apollo-android/blob/master/apollo-normalized-cache-sqlite/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/sql/SqlNormalizedCache.kt#L56

```
cacheQueries.update(oldRecord.key, recordFieldAdapter.toJson(oldRecord.fields))
```

but it should be
```
cacheQueries.update(recordFieldAdapter.toJson(oldRecord.fields), oldRecord.key)
or even safer
cacheQueries.update(record = recordFieldAdapter.toJson(oldRecord.fields), key = oldRecord.key)
```

as it generated from the 
```
update:
UPDATE records SET record=:record WHERE key=:key;
```

**Comment by sav007** [comment]
So the issue is not in SQL DML but with the order of passed arguments, I opened new PR https://github.com/apollographql/apollo-android/pull/2245 that fixes it.

**Comment by PierLucaLW** [comment]
@tasomaniac @sav007 ooh I see, haven't thought about that at all. Alright will just close this one then, thanks for looking into it.

**Pull request closed without merging**